Client/DataAccess/SQL/JoinedHistoryVisitsTable.swift

```diff
@@ -0,0 +1,117 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+let HistoryVisits = "history-visits"
+
+class JoinedHistoryVisitsTable: Table {
+    var name: String { return HistoryVisits }
+
+    private let visits = VisitsTable<Visit>()
+    private let history = HistoryTable<Site>()
+
+    private func getGuidFor(db: SQLiteDBConnection, site: Site) -> String? {
+        let opts = QueryOptions()
+        opts.filter = site.url
+
+        let cursor = history.query(db, options: opts)
+        if (cursor.count != 1) {
+            return nil
+        }
+        return (cursor[0] as Site).guid
+    }
+
+    func create(db: SQLiteDBConnection, version: Int) -> Bool {
+        return history.create(db, version: version) && visits.create(db, version: version)
+    }
+
+    func updateTable(db: SQLiteDBConnection, from: Int, to: Int) -> Bool {
+        return history.updateTable(db, from: from, to: to) && visits.updateTable(db, from: from, to: to)
+    }
+
+    private func updateSite(db: SQLiteDBConnection, site: Site, inout err: NSError?) -> Int {
+        // If our site doesn't have a guid, we need to find one
+        if site.guid == nil {
+            if let guid = getGuidFor(db, site: site) {
+                site.guid = guid
+                // Update the page title
+                return history.update(db, item: site, err: &err);
+            } else {
+                // Make sure we have a site in the table first
+                site.guid = NSUUID().UUIDString
+                return history.insert(db, item: site, err: &err)
+            }
+        }
+
+        // Update the page title
+        return history.update(db, item: site, err: &err)
+    }
+
+    func insert<T>(db: SQLiteDBConnection, item: T?, inout err: NSError?) -> Int {
+        if let visit = item as? Visit {
+            if updateSite(db, site: visit.site, err: &err) < 0 {
+                return -1;
+            }
+
+            // Now add a visit
+            return visits.insert(db, item: visit, err: &err)
+        } else if let site = item as? Site {
```

Comment:
Yeah, this runs into a Swift problem that I haven't found a good solution to yet (its also why our Cursors aren't generic). i.e. https://stackoverflow.com/questions/24888560/usage-of-protocols-as-array-types-and-function-parameters-in-swift

Basically, if you make a protocol generic, it isn't useful anymore outside of a constraint on a class definition. You can declare a class that implements the protocol Table<T>, but you can't store an array of Table<T>[] anymore. I've tried to get around this a few times. For instance, declaring base-class that implements the protocol and having everyone inherit from that, but even then storing an array of BaseTable<T> where T varies from obj to obj isn't possible (AFAICT). That's why I've just taken to passing around a lot of Any types here. Its as clean as I can seem to make things...

An alternative idea I had: remove the array of tables entirely from BrowserDB and do things more like we are in tests? i.e.

BrowserDB.insert(tableName, object) would become:
BrowserDB.withConnection(.ReadWrite) { conn in table.insert(conn, object) }

Let me give that a shot.